google/resource_dataproc_job.go

```diff
@@ -0,0 +1,840 @@
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+
+				"main_class": {
+					Type:          schema.TypeString,
+					Optional:      true,
+					ForceNew:      true,
+					ConflictsWith: []string{"spark_config.main_jar"},
+				},
+
+				"main_jar": {
+					Type:          schema.TypeString,
+					Optional:      true,
+					ForceNew:      true,
+					ConflictsWith: []string{"spark_config.main_class"},
+				},
+
+				"jars": {
+					Type:     schema.TypeList,
+					Optional: true,
+					ForceNew: true,
+					Elem:     &schema.Schema{Type: schema.TypeString},
+				},
+
+				"files": {
+					Type:     schema.TypeList,
+					Optional: true,
+					ForceNew: true,
+					Elem:     &schema.Schema{Type: schema.TypeString},
+				},
+
+				"archives": {
+					Type:     schema.TypeList,
+					Optional: true,
+					ForceNew: true,
+					Elem:     &schema.Schema{Type: schema.TypeString},
+				},
+
+				"args": {
+					Type:     schema.TypeList,
+					Optional: true,
+					ForceNew: true,
+					Elem:     &schema.Schema{Type: schema.TypeString},
+				},
+
+				"properties": {
+					Type:     schema.TypeMap,
+					Optional: true,
+					ForceNew: true,
+					Elem:     &schema.Schema{Type: schema.TypeString},
+				},
+			},
+		},
+	}
+}
+
+func getTfSparkConfig(job *dataproc.SparkJob) []map[string]interface{} {
+	return []map[string]interface{}{
+		{
+			"main_class": job.MainClass,
+			"main_jar":   job.MainJarFileUri,
+			"jars":       job.JarFileUris,
+			"files":      job.FileUris,
+			"archives":   job.ArchiveUris,
+			"args":       job.Args,
+			"properties": job.Properties,
+		},
+	}
+}
+
+func getSparkJob(config map[string]interface{}) *dataproc.SparkJob {
+
+	job := &dataproc.SparkJob{}
+	if v, ok := config["main_class"]; ok {
+		job.MainClass = v.(string)
+	}
+	if v, ok := config["main_jar"]; ok {
+		job.MainJarFileUri = v.(string)
+	}
+
+	if v, ok := config["args"]; ok {
+		job.Args = convertStringArr(v.([]interface{}))
+	}
+	if v, ok := config["properties"]; ok {
+		job.Properties = convertMapString(v.(map[string]interface{}))
+	}
+	if v, ok := config["jars"]; ok {
+		job.JarFileUris = convertStringArr(v.([]interface{}))
+	}
+	if v, ok := config["archives"]; ok {
+		job.ArchiveUris = convertStringArr(v.([]interface{}))
+	}
+	if v, ok := config["files"]; ok {
+		job.FileUris = convertStringArr(v.([]interface{}))
+	}
+
+	return job
+
+}
+
+// ---- Hadoop Job ----
+
+func hadoopTFSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:          schema.TypeList,
+		Optional:      true,
+		ForceNew:      true,
+		MaxItems:      1,
+		ConflictsWith: []string{"spark_config", "pyspark_config", "hive_config", "pig_config", "sparksql_config"},
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+
+				"main_class": {
+					Type:          schema.TypeString,
+					Optional:      true,
+					ForceNew:      true,
+					ConflictsWith: []string{"hadoop_config.main_jar"},
```

Comment:
because of the way lists are represented in state, this should (I believe) be `hadoop_config.0.main_jar`